File: apps/android/app/src/main/java/com/dndsync/android/sync/LanSyncGate.kt
package com.dndsync.android.sync

import com.dndsync.proto.v1.DndState

class LanSyncGate(
    private val nowMs: () -> Long = { System.currentTimeMillis() },
) {
    var onOriginate: ((DndState) -> Unit)? = null
    var onApplyRemote: ((Boolean) -> Unit)? = null

    private val lock = Any()
    private var lastAppliedUnixMs: Long = 0
    private var lastAppliedOn: Boolean? = null
    private var applyingRemote = false
    private var echoUntilMs: Long = 0

    fun noteObserverEvent() {
        synchronized(lock) {
            applyingRemote = false
        }
    }

    fun reset() {
        synchronized(lock) {
            lastAppliedUnixMs = 0
            lastAppliedOn = null
            applyingRemote = false
            echoUntilMs = 0
        }
    }

    fun onLocalChange(on: Boolean) {
        val msg = synchronized(lock) {
            val now = nowMs()
            if (now >= echoUntilMs) {
                applyingRemote = false
            }
            if (applyingRemote || now < echoUntilMs) {
                return
            }
            // Applying a remote on/off can take longer than the 1s echo
            // window (Mac Shortcuts especially). The observer that then
            // fires is that apply, not a new user action.
            if (lastAppliedOn == on) {
                return
            }
            lastAppliedUnixMs = now
            lastAppliedOn = on
            DndState.newBuilder()
                .setVersion(Wire.PROTO_VERSION)
                .setOn(on)
                .setUnixMs(now)
                .setSender(Wire.SENDER_ANDROID)
                .build()
        }
        onOriginate?.invoke(msg)
    }

    fun onRemote(msg: DndState): Boolean {
        if (msg.version != Wire.PROTO_VERSION) {
            return false
        }
        val on = synchronized(lock) {
            if (msg.unixMs <= lastAppliedUnixMs) {
                return false
            }
            lastAppliedUnixMs = msg.unixMs
            lastAppliedOn = msg.on
            applyingRemote = true
            echoUntilMs = nowMs() + Wire.ECHO_WINDOW_MS
            msg.on
        }
        onApplyRemote?.invoke(on)
        return true
    }
}

File: apps/macos/FocusSync/LanSyncGate.swift
import Foundation

final class LanSyncGate {
    var onOriginate: ((Dndsync_V1_DndState) -> Void)?
    var onApplyRemote: ((Bool) -> Void)?

    private let sender: String
    private let nowMs: () -> Int64
    private var lastAppliedUnixMs: Int64 = 0
    private var lastAppliedOn: Bool?
    private var applyingRemote = false
    private var echoUntilMs: Int64 = 0

    init(sender: String, nowMs: @escaping () -> Int64 = {
        Int64(Date().timeIntervalSince1970 * 1000)
    }) {
        self.sender = sender
        self.nowMs = nowMs
    }

    func noteObserverEvent() {
        applyingRemote = false
    }

    func reset() {
        lastAppliedUnixMs = 0
        lastAppliedOn = nil
        applyingRemote = false
        echoUntilMs = 0
    }

    func onLocalChange(on: Bool) {
        let now = nowMs()
        if now >= echoUntilMs {
            applyingRemote = false
        }
        if applyingRemote || now < echoUntilMs {
            return
        }
        // Shortcuts can take longer than the 1s echo window to flip Focus.
        // The observer that then fires is the apply we just did, not a new
        // user action — originating it would log a duplicate "by this Mac"
        // row and bounce the same on/off back to the phone.
        if lastAppliedOn == on {
            return
        }
        let msg = DndStateFrames.make(on: on, unixMs: now, sender: sender)
        lastAppliedUnixMs = msg.unixMs
        lastAppliedOn = on
        onOriginate?(msg)
    }

    func onRemote(_ msg: Dndsync_V1_DndState) -> Bool {
        if msg.version != LanConstants.protoVersion {
            return false
        }
        if msg.unixMs <= lastAppliedUnixMs {
            return false
        }
        lastAppliedUnixMs = msg.unixMs
        lastAppliedOn = msg.on
        applyingRemote = true
        echoUntilMs = nowMs() + LanConstants.echoWindowMs
        onApplyRemote?(msg.on)
        return true
    }
}
